Use redefining helper pattern for better tree-shakeability

Andarist opened the pull request:
* [x] Code is up-to-date with the `master` branch
* [x] You've successfully run `jake runtests` locally
* [x] You've signed the CLA

I'm not yet sure if I've made this change correctly, need to fix tests first - but this presents the idea, so let me know if you'd like to proceed with this or not. By changing `extendStatics` & `assign` to declarations they become effectively more static and thus it makes the whole thing more recognizable as side-effect free, so other tooling has easier job at tree-shaking them.

We use this pattern currently in babel. You can check [here](https://github.com/babel/babel/blob/25153359f7ce5b52a56d1911089addcf43ae8cf8/packages/babel-helpers/src/helpers.js#L359).

~~This PR fails at:~~
> ~~TypeError: Assignment to constant variable.~~

~~I suspect that those helpers are transformed to `const` somewhere, but I don't know where yet. I'd appreciate some hint about this one~~

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/TypeScript?pullRequest=24244) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/TypeScript?pullRequest=24244) before we can accept your contribution.<br/><br/>:x: Andarist [sign now](https://cla.opensource.microsoft.com/Microsoft/TypeScript?pullRequest=24244)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/TypeScript?pullRequest=24244) it.</sub>

j-oliveras commented:
Your change is not valid. Running your `__assign` version with node 8.11.1 calling as `__assign({}, {a: 1})` I get an `RangeError: Maximum call stack size exceeded` exception instead of `{ a: 1}`.

Andarist commented:
You are right, I've reassigned wrong identifier - it should be fixed now.

mhegazy commented:
@rbuckton thoughts?

rbuckton commented:
@mhegazy - These changes look fine to me. 

@Andarist - Where are you seeing "TypeError: Assignment to constant variable."? Is it from reassigning `extendsStatics` since it is declared as a function?

Andarist commented:
> @Andarist - Where are you seeing "TypeError: Assignment to constant variable."? Is it from reassigning extendsStatics since it is declared as a function?

Not sure, but I was reassigning wrong identifier at first - I've fixed it and the problem is gone. However still the received error seemed to be wrong as my code error was causing runtime recursion error and not `TypeError`, but 🤷‍♂️ 

I've updated baseline tests, from my perspective it is ready to merge if you decide the change is worth it 😉

j-oliveras commented:
I think that the @rbuckton comment is because your current `__extends` [at playground](https://www.typescriptlang.org/play/#src=var%20__extends%20%3D%20(this%20%26%26%20this.__extends)%20%7C%7C%20(function%20()%20%7B%0D%0A%09function%20extendStatics(d%2C%20b)%20%7B%0D%0A%09%09extendStatics%20%3D%20Object.setPrototypeOf%20%7C%7C%0D%0A%09%09%09(%7B%20__proto__%3A%20%5B%5D%20%7D%20instanceof%20Array%20%26%26%20function%20(d%2C%20b)%20%7B%20d.__proto__%20%3D%20b%3B%20%7D)%20%7C%7C%0D%0A%09%09%09function%20(d%2C%20b)%20%7B%20for%20(var%20p%20in%20b)%20if%20(b.hasOwnProperty(p))%20d%5Bp%5D%20%3D%20b%5Bp%5D%3B%20%7D%3B%0D%0A%09%09return%20extendStatics(d%2C%20b)%3B%0D%0A%09%7D%0D%0A%0D%0A%09return%20function%20(d%2C%20b)%20%7B%0D%0A%09%09extendStatics(d%2C%20b)%3B%0D%0A%09%09function%20__()%20%7B%20this.constructor%20%3D%20d%3B%20%7D%0D%0A%09%09d.prototype%20%3D%20b%20%3D%3D%3D%20null%20%3F%20Object.create(b)%20%3A%20(__.prototype%20%3D%20b.prototype%2C%20new%20__())%3B%0D%0A%09%7D%3B%0D%0A%7D)()%3B%0D%0A) has a compile error at `extendStatics` assignment with message `Cannot assign to 'extendStatics' because it is not a variable.` The code is:
```ts
var __extends = (this && this.__extends) || (function () {
	function extendStatics(d, b) {
		extendStatics = Object.setPrototypeOf ||
			({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
			function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
		return extendStatics(d, b);
	}

	return function (d, b) {
		extendStatics(d, b);
		function __() { this.constructor = d; }
		d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
	};
})();
```
Seems it is valid to reassign a function in javascript but typescript don't allow it. I don't know if is by design.

How to change it to this?
```ts
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
```

Andarist commented:
Sure, I can change `extendStatics` to function expression if needed - just let me know.

Andarist commented:
@rbuckton friendly 🏓 is there anything left to do here? is this something that you would like to incorporate?

rbuckton commented on src/compiler/transformers/es2015.ts:
```diff
@@ -4074,9 +4074,13 @@ namespace ts {
-                var extendStatics = Object.setPrototypeOf ||
-                    ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
-                    function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
+                function extendStatics(d, b) {
```
Could you make this a `var` per https://github.com/Microsoft/TypeScript/pull/24244#issuecomment-390468895?

Andarist commented:
> One minor change, per the earlier discussion.

Done.

> Also, would you be interested in making this change in https://github.com/Microsoft/tslib as well?

That was my original intention/reason why I've started this discussion here 😉 Is `tslib` manually updated or is it done with some script?

mhegazy commented:
@rbuckton can u take one more look?

Andarist commented:
@rbuckton friendly 🏓 😉

rbuckton commented:
Sorry for the delay. 

> Is tslib manually updated or is it done with some script?

tslib is manually updated currently.

Andarist commented:
> tslib is manually updated currently.

Ok, thanks for the info - gonna prepare a followup PR there.
